From 820dc1ebf2911eda341335edba5b66dae5b3bf9b Mon Sep 17 00:00:00 2001
From: Brecht Van Lommel <brecht@blender.org>
Date: Thu, 23 Oct 2025 20:18:11 +0200
Subject: [PATCH 06/22] BLENDER: Projects: button to show/hide issue details
 and closed issue

Both are off by default. This is implemented fully on the frontend, so
all issues and their details are still always loaded.
---
 modules/templates/helper.go       | 18 ++++++++-----
 templates/projects/view.tmpl      | 43 ++++++++++++++++++++++++++++++-
 templates/repo/issue/card.tmpl    | 27 +++++++++----------
 web_src/css/features/projects.css |  8 ++++++
 web_src/css/repo/issue-card.css   |  2 ++
 5 files changed, 78 insertions(+), 20 deletions(-)

diff --git a/modules/templates/helper.go b/modules/templates/helper.go
index 662b89998e..f169c2ed73 100644
--- a/modules/templates/helper.go
+++ b/modules/templates/helper.go
@@ -30,12 +30,13 @@ func newFuncMapWebPage() template.FuncMap {
 
 		// -----------------------------------------------------------------
 		// html/template related functions
-		"dict":        dict, // it's lowercase because this name has been widely used. Our other functions should have uppercase names.
-		"Iif":         iif,
-		"Eval":        evalTokens,
-		"HTMLFormat":  htmlFormat,
-		"QueryEscape": queryEscape,
-		"QueryBuild":  QueryBuild,
+		"dict":        dict, // it's lowercase because this name has been widely used. Our other functions should have uppercase names.
+		"Iif":         iif,
+		"Eval":        evalTokens,
+		"HTMLFormat":  htmlFormat,
+		"QueryEscape": queryEscape,
+		"QueryBuild":  QueryBuild,
+		"JSEscape":    jsEscapeSafe,
 
 		"PathEscape":         url.PathEscape,
 		"PathEscapeSegments": util.PathEscapeSegments,
@@ -160,6 +162,10 @@ func htmlFormat(s any, args ...any) template.HTML {
 	panic(fmt.Sprintf("unexpected type %T", s))
 }
 
+func jsEscapeSafe(s string) template.HTML {
+	return template.HTML(template.JSEscapeString(s))
+}
+
 func queryEscape(s string) template.URL {
 	return template.URL(url.QueryEscape(s))
 }
diff --git a/templates/projects/view.tmpl b/templates/projects/view.tmpl
index 8f4f1d60fd..f8b969fa2b 100644
--- a/templates/projects/view.tmpl
+++ b/templates/projects/view.tmpl
@@ -23,6 +23,16 @@
 				"ClosedMilestones" .ClosedMilestones
 			}}
 		</div>
+			<div class="ui compact mini menu" style="margin-left: auto; margin-right: 1em;">
+				<button class="item btn" onclick="clickDetailsView()">
+					<span class="issue-card-details">Show Details</span>
+					<span class="issue-card-details tw-hidden">Hide Details</span>
+				</button>
+				<button class="item btn" onclick="clickClosedView()">
+					<span class="issue-card-closed">Show Closed</span>
+					<span class="issue-card-closed tw-hidden">Hide Closed</span>
+				</button>
+			</div>
 		{{if $canWriteProject}}
 			<div class="ui compact mini menu">
 				<a class="item screen-full">
@@ -122,7 +132,7 @@
 				<div class="divider"{{if .Color}} style="color: {{ContrastColor .Color}} !important"{{end}}></div>
 				<div class="ui cards" data-url="{{$.Link}}/{{.ID}}" data-project="{{$.Project.ID}}" data-board="{{.ID}}" id="board_{{.ID}}">
 					{{range (index $.IssuesMap .ID)}}
-						<div class="issue-card tw-break-anywhere {{if $canWriteProject}}tw-cursor-grab{{end}}" data-issue="{{.ID}}">
+						<div class="issue-card tw-break-anywhere {{if $canWriteProject}}tw-cursor-grab{{end}} {{if .IsClosed}}issue-card-closed tw-hidden{{end}}" data-issue="{{.ID}}">
 							{{template "repo/issue/card" (dict "Issue" . "Page" $)}}
 						</div>
 					{{end}}
@@ -132,6 +142,37 @@
 	</div>
 </div>
 
+<script>
+	keyDetails = 'project-show-issue-details:{{.Repository.FullName | JSEscape}}:{{.Project.ID}}';
+	keyClosed = 'project-show-issue-closed:{{.Repository.FullName | JSEscape}}:{{.Project.ID}}';
+
+	function toggleDetailsView() {
+		for (const elem of document.querySelectorAll('.issue-card-details')) {
+			elem.classList.toggle('tw-hidden');
+		}
+	}
+	function clickDetailsView() {
+		localStorage.setItem(keyDetails, (localStorage.getItem(keyDetails) === 'true') ? 'false' : 'true');
+		toggleDetailsView();
+	}
+	if (localStorage.getItem(keyDetails) === 'true') {
+		toggleDetailsView();
+	}
+
+	function toggleClosedView() {
+		for (const elem of document.querySelectorAll('.issue-card-closed')) {
+			elem.classList.toggle('tw-hidden');
+		}
+	}
+	function clickClosedView() {
+		localStorage.setItem(keyClosed, (localStorage.getItem(keyClosed) === 'true') ? 'false' : 'true');
+		toggleClosedView();
+	}
+	if (localStorage.getItem(keyClosed) === 'true') {
+		toggleClosedView();
+	}
+</script>
+
 {{if $canWriteProject}}
 <div class="ui small modal" id="project-column-modal-edit">
 	<div class="header">{{ctx.Locale.Tr "repo.projects.column.edit"}}</div>
diff --git a/templates/repo/issue/card.tmpl b/templates/repo/issue/card.tmpl
index 3c3a1dc047..7834e8b4dc 100644
--- a/templates/repo/issue/card.tmpl
+++ b/templates/repo/issue/card.tmpl
@@ -10,6 +10,14 @@
 		{{end}}
 	{{end}}
 	<div class="content tw-w-full">
+		<!-- BLENDER: Assignees here for more compact display -->
+		{{if .Assignees}}
+		<div class="issue-card-assignees">
+			{{range .Assignees}}
+				<a target="_blank" href="{{.HomeLink}}" data-tooltip-content="{{ctx.Locale.Tr "repo.projects.column.assigned_to"}} {{.Name}}">{{ctx.AvatarUtils.Avatar . 28 "mini gt-mr-3"}}</a>
+			{{end}}
+		</div>
+		{{end}}
 		<div class="tw-flex tw-items-start tw-gap-[5px]">
 			<div class="issue-card-icon">
 				{{template "shared/issueicon" .}}
@@ -22,7 +30,7 @@
 			{{end}}
 		</div>
 		<div class="meta">
-			<span class="tw-text-grey-light muted-links">
+			<span class="tw-text-grey-light muted-links issue-card-details tw-hidden">
 				{{if not $.Page.Repository}}{{.Repo.FullName}}{{end}}#{{.Index}}
 				{{$timeStr := DateUtils.TimeSince .GetLastEventTimestamp}}
 				{{if .OriginalAuthor}}
@@ -35,7 +43,7 @@
 			</span>
 		</div>
 		{{if .MilestoneID}}
-		<div class="meta tw-my-1">
+		<div class="meta tw-my-1 issue-card-details tw-hidden">
 			<a class="milestone" href="{{.Repo.Link}}/milestone/{{.MilestoneID}}">
 				{{svg "octicon-milestone" 16 "tw-mr-1 tw-align-middle"}}
 				<span class="tw-align-middle">{{.Milestone.Name}}</span>
@@ -44,7 +52,7 @@
 		{{end}}
 		{{if $.Page.LinkedPRs}}
 		{{range index $.Page.LinkedPRs .ID}}
-		<div class="meta tw-my-1">
+		<div class="meta tw-my-1 issue-card-details tw-hidden">
 			<a href="{{.Repo.Link}}/pulls/{{.Index}}">
 				<span class="tw-m-0 {{if .PullRequest.HasMerged}}tw-text-purple{{else if .IsClosed}}tw-text-red{{else}}tw-text-green{{end}}">{{svg "octicon-git-merge" 16 "tw-mr-1 tw-align-middle"}}</span>
 				<span class="tw-align-middle">{{.Title}} <span class="tw-text-grey-light">#{{.Index}}</span></span>
@@ -54,15 +62,15 @@
 		{{end}}
 		{{$tasks := .GetTasks}}
 		{{if gt $tasks 0}}
-			<div class="meta tw-my-1">
+			<div class="meta tw-my-1 issue-card-details tw-hidden">
 				{{svg "octicon-checklist" 16 "tw-mr-1 tw-align-middle"}}
 				<span class="tw-align-middle">{{.GetTasksDone}} / {{$tasks}}</span>
 			</div>
 		{{end}}
 	</div>
 
-	{{if or .Labels .Assignees}}
-	<div class="issue-card-bottom">
+	{{if or .Labels}}
+	<div class="issue-card-bottom issue-card-details tw-hidden">
 		{{/* the labels container must always be present, to help the assignees list right-aligned */}}
 		<div class="issue-card-bottom-part labels-list">
 			{{range $label := .Labels}}
@@ -71,13 +79,6 @@
 				<a class="item" href="{{$link}}">{{ctx.RenderUtils.RenderLabel $label}}</a>
 			{{end}}
 		</div>
-		{{if .Assignees}}
-		<div class="issue-card-bottom-part tw-justify-end">
-			{{range .Assignees}}
-				<a target="_blank" href="{{.HomeLink}}" data-tooltip-content="{{ctx.Locale.Tr "repo.projects.column.assigned_to"}} {{.Name}}">{{ctx.AvatarUtils.Avatar . 24}}</a>
-			{{end}}
-		</div>
-		{{end}}
 	</div>
 	{{end}}
 {{end}}
diff --git a/web_src/css/features/projects.css b/web_src/css/features/projects.css
index 25cb530f85..bfffa31138 100644
--- a/web_src/css/features/projects.css
+++ b/web_src/css/features/projects.css
@@ -20,6 +20,14 @@
 }
 
 .project-column {
+  background-color: var(--color-project-column-bg) !important;
+  border: 1px solid var(--color-secondary) !important;
+  border-radius: var(--border-radius);
+  margin: 0 0.5rem !important;
+  padding: 0.5rem !important;
+  width: 320px;
+  height: initial;
+  min-height: max(calc(100vh - 225px), 300px);
   flex: 0 0 auto;
   display: flex;
   flex-direction: column;
diff --git a/web_src/css/repo/issue-card.css b/web_src/css/repo/issue-card.css
index 327919b1fe..1067b5c8f7 100644
--- a/web_src/css/repo/issue-card.css
+++ b/web_src/css/repo/issue-card.css
@@ -41,4 +41,6 @@
   overflow: hidden;
   max-width: fit-content;
   max-height: fit-content;
+  float: right;
+  margin-left: 0.5em;
 }
-- 
2.53.0

